### PR TITLE
Added RBAC ports to loadbalancer preparation

### PR DESCRIPTION
**Pull request by c3y1huang**
Caused by: https://github.com/SUSE/avant-garde/issues/475

Added RBAC authentication ports. This is required to handle web-server redirection between Gangway and Dex.

Signed-off-by: Chin-Ya Huang <chin-ya.huang@suse.com>

### CODE REVIEW

File: adoc/deployment-preparation.adoc [hunk 1]
```diff
@@ -116,7 +116,8 @@ during the cluster bootstrap. To simplify configuration you can reserve the IPs
 needed for the cluster nodes and pre-configure these in the load balancer.
 
 The load balancer needs access to port `6443` on the `apiserver` (all master nodes)
-in the cluster.
+in the cluster. It also needs access to Gangway port `32001` and Dex port `32002` 
+on all worker nodes in the cluster for RBAC authentication.
```
Comment (jhsiaosuse) [inline adoc/deployment-preparation.adoc]: Reference [v3](https://www.suse.com/documentation/suse-caasp-3/book_caasp_deployment/data/sec_deploy_requirements_network.html), both masters and workers expose ports.

image

Comment (c3y1huang) [reply adoc/deployment-preparation.adoc]: Fixed, thanks :)

**Comment by Martin-Weiss** [comment]
> Am 09.07.2019 um 04:35 schrieb c3y1huang <notifications@github.com>:
> 
> @c3y1huang commented on this pull request.
> 
> In adoc/deployment-preparation.adoc:
> 
> > @@ -116,7 +116,8 @@ during the cluster bootstrap. To simplify configuration you can reserve the IPs
>  needed for the cluster nodes and pre-configure these in the load balancer.
>  
>  The load balancer needs access to port `6443` on the `apiserver` (all master nodes)
> -in the cluster.
> +in the cluster. It also needs access to Gangway port `32001` and Dex port `32002` 
> +on all worker nodes in the cluster for RBAC authentication.
> Fixed, thanks :)
> 

Is there a reason why we shift from 32000 to 32002 for Dex?

**Comment by jhsiaosuse** [comment]
> Am 09.07.2019 um 04:35 schrieb c3y1huang ***@***.***>: @c3y1huang commented on this pull request. In adoc/deployment-preparation.adoc: > @@ -116,7 +116,8 @@ during the cluster bootstrap. To simplify configuration you can reserve the IPs needed for the cluster nodes and pre-configure these in the load balancer. The load balancer needs access to port `6443` on the `apiserver` (all master nodes) -in the cluster. +in the cluster. It also needs access to Gangway port `32001` and Dex port `32002` +on all worker nodes in the cluster for RBAC authentication. Fixed, thanks :)
> Is there a reason why we shift from 32000 to 32002 for Dex?
> […](#)
> — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub, or mute the thread.



> Am 09.07.2019 um 04:35 schrieb c3y1huang ***@***.***>: @c3y1huang commented on this pull request. In adoc/deployment-preparation.adoc: > @@ -116,7 +116,8 @@ during the cluster bootstrap. To simplify configuration you can reserve the IPs needed for the cluster nodes and pre-configure these in the load balancer. The load balancer needs access to port `6443` on the `apiserver` (all master nodes) -in the cluster. +in the cluster. It also needs access to Gangway port `32001` and Dex port `32002` +on all worker nodes in the cluster for RBAC authentication. Fixed, thanks :)
> Is there a reason why we shift from 32000 to 32002 for Dex?
> […](#)
> — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub, or mute the thread.

No real reason actually, it should not have any forward/backward compatible issue, right?

**Comment by r0ckarong** [comment]
@Martin-Weiss @jhsiaosuse Is the discussion about the Dex port solved? I can not merge with an ongoing debate.

**Comment by innobead** [comment]
@Martin-Weiss please feedback if u have concerns after @jhsiaosuse answer, otherwise we need to go merge today for beta 4 release. Thanks.

**Comment by c3y1huang** [comment]
Updated dex port to 32000. @jhsiaosuse please assist with port change.

**Comment by jhsiaosuse** [comment]
> Updated dex port to 32000. @jhsiaosuse please assist with port change.

@Martin-Weiss @r0ckarong 
Change dex port to `32000`, [PR](https://github.com/SUSE/skuba/pull/464)

**Comment by Martin-Weiss** [comment]
> 
> 
> > Updated dex port to 32000. @jhsiaosuse please assist with port change.
> 
> @Martin-Weiss @r0ckarong
> Change dex port back to `32000` (align with v3), [PR](https://github.com/SUSE/skuba/pull/464)

Thanks so much!! IMO it is important to keep 32000 as in previous CaaSP versions we had this port and with the manual configuration of load balancers at the customer side for 6443 and 32000 - keeping 32000 at least allows to keep the existing load  balancer configuration for the upgrade from 3 to 4..  (If I understand the architecture for Dex right - I am not sure what / who will need access to gangway which way..)

**Comment by jhsiaosuse** [comment]
> > > Updated dex port to 32000. @jhsiaosuse please assist with port change.
> > 
> > 
> > @Martin-Weiss @r0ckarong
> > Change dex port back to `32000` (align with v3), [PR](https://github.com/SUSE/skuba/pull/464)
> 
> Thanks so much!! IMO it is important to keep 32000 as in previous CaaSP versions we had this port and with the manual configuration of load balancers at the customer side for 6443 and 32000 - keeping 32000 at least allows to keep the existing load balancer configuration for the upgrade from 3 to 4.. (If I understand the architecture for Dex right - I am not sure what / who will need access to gangway which way..)

Every k8s users who wanna access cluster with authentication and RBAC, they need to access gangway first and then gangway will redirect them to dex for authentication, after authenticated, dex will redirect back to gangway for let user get KUBECONFIG.

This different as v3. For v3, authentication and RBAC only needs dex to get KUBECONFIG; but for v4, authentication need dex + gangway together to get KUBECONFIG.

So both port `32000` and `32001` need to config on customer's load balancer.

**Comment by Martin-Weiss** [comment]
> 
> 
> > > > Updated dex port to 32000. @jhsiaosuse please assist with port change.
> > > 
> > > 
> > > @Martin-Weiss @r0ckarong
> > > Change dex port back to `32000` (align with v3), [PR](https://github.com/SUSE/skuba/pull/464)
> > 
> > 
> > Thanks so much!! IMO it is important to keep 32000 as in previous CaaSP versions we had this port and with the manual configuration of load balancers at the customer side for 6443 and 32000 - keeping 32000 at least allows to keep the existing load balancer configuration for the upgrade from 3 to 4.. (If I understand the architecture for Dex right - I am not sure what / who will need access to gangway which way..)
> 
> Every k8s users who wanna access cluster with authentication and RBAC, they need to access gangway first and then gangway will redirect them to dex for authentication, after authenticated, dex will redirect back to gangway for let user get KUBECONFIG.
> 
> This different as v3. For v3, authentication and RBAC only needs dex to get KUBECONFIG; but for v4, authentication need dex + gangway together to get KUBECONFIG.
> 
> So both port `32000` and `32001` need to config on customer's load balancer.

Understood - so 6443+32000+32001 need to be on the VIP of the load balancer to the masters?

Just a quick ides - could we put gangway and dex "behind" nginx-ingress and run nginx-ingress on 32000?

**Comment by innobead** [comment]
> > > > > Updated dex port to 32000. @jhsiaosuse please assist with port change.
> > > > 
> > > > 
> > > > @Martin-Weiss @r0ckarong
> > > > Change dex port back to `32000` (align with v3), [PR](https://github.com/SUSE/skuba/pull/464)
> > > 
> > > 
> > > Thanks so much!! IMO it is important to keep 32000 as in previous CaaSP versions we had this port and with the manual configuration of load balancers at the customer side for 6443 and 32000 - keeping 32000 at least allows to keep the existing load balancer configuration for the upgrade from 3 to 4.. (If I understand the architecture for Dex right - I am not sure what / who will need access to gangway which way..)
> > 
> > 
> > Every k8s users who wanna access cluster with authentication and RBAC, they need to access gangway first and then gangway will redirect them to dex for authentication, after authenticated, dex will redirect back to gangway for let user get KUBECONFIG.
> > This different as v3. For v3, authentication and RBAC only needs dex to get KUBECONFIG; but for v4, authentication need dex + gangway together to get KUBECONFIG.
> > So both port `32000` and `32001` need to config on customer's load balancer.
> 
> Understood - so 6443+32000+32001 need to be on the VIP of the load balancer to the masters?
> 
> Just a quick idea - could we put gangway and dex "behind" nginx-ingress and run nginx-ingress on 32000?

Technically yes, but this needs to update gangway & dex configurations to make this work. We would suggest best practice first.

**Comment by Martin-Weiss** [comment]
> > > So both port `32000` and `32001` need to config on customer's load balancer.
> > 
> > 
> > Understood - so 6443+32000+32001 need to be on the VIP of the load balancer to the masters?
> > Just a quick idea - could we put gangway and dex "behind" nginx-ingress and run nginx-ingress on 32000?
> 
> Technically yes, but this needs to update gangway & dex configurations to make this work. We would suggest best practice first. Let's discuss later if have more question/idea after beta 4 out :)

File - lets start with 32000+32001 - and not make it too complex ;-)

**Comment by r0ckarong** [comment]
@innobead @c3y1huang Would you please amend the ports list too:

https://github.com/SUSE/doc-caasp/blame/master/adoc/deployment-sysreqs.adoc#L95

**Comment by innobead** [comment]
@c3y1huang please help on that, thanks.

**Comment by c3y1huang** [comment]
Thanks @r0ckarong :)

**Pull request merged**